cvat/apps/engine/serializers.py

```diff
@@ -1980,63 +1982,64 @@ class Meta:
         read_only_fields = ('path',)
 
 
-def _update_related_storages(instance, validated_data):
-    for storage in ('source_storage', 'target_storage'):
-        new_conf = validated_data.pop(storage, None)
+def _update_related_storages(
+    instance: Union[models.Project, models.Task],
+    validated_data: Dict[str, Any],
+) -> None:
+    for storage_type in ('source_storage', 'target_storage'):
+        new_conf = validated_data.pop(storage_type, None)
 
         if not new_conf:
             continue
 
+        cloud_storage = None
         cloud_storage_id = new_conf.get('cloud_storage_id')
-        if cloud_storage_id:
-            _validate_existence_of_cloud_storage(cloud_storage_id)
+        new_location = new_conf.get('location')
 
         # storage_instance maybe None
-        storage_instance = getattr(instance, storage)
-        if not storage_instance:
-            storage_instance = models.Storage(**new_conf)
-            storage_instance.save()
-            setattr(instance, storage, storage_instance)
-            continue
+        storage_instance = getattr(instance, storage_type)
 
-        new_location = new_conf.get('location')
-        storage_instance.location = new_location or storage_instance.location
-        storage_instance.cloud_storage_id = new_conf.get('cloud_storage_id', \
-            storage_instance.cloud_storage_id if not new_location else None)
+        if not cloud_storage_id and new_location == models.Location.CLOUD_STORAGE:
+            raise serializers.ValidationError('Cloud storage was selected as location but its id was not specified')
+        elif cloud_storage_id and new_location and new_location != models.Location.CLOUD_STORAGE:
+            raise serializers.ValidationError(f"It is not allowed to specify '{new_location}' location together with cloud storage id")
+        elif cloud_storage_id and not new_location and getattr(storage_instance, 'location', None) != models.Location.CLOUD_STORAGE:
+            raise serializers.ValidationError(f"The configuration of {storage_type} is not full")
 
-        cloud_storage_id = storage_instance.cloud_storage_id
         if cloud_storage_id:
             try:
-                _ = models.CloudStorage.objects.get(id=cloud_storage_id)
+                cloud_storage = models.CloudStorage.objects.get(id=cloud_storage_id)
```

Comment:
I don't think this variable is actually needed, since you can use `cloud_storage_id` when updating the `Storage` object (and you already do that when you create a new `Storage` object, so might as well be consistent).